### ISSUE TITLE
All notification functionality

### ISSUE THREAD

**Issue by OddArneSaetervik** [Altinn/altinn-notifications#60]
## Description
|This issue will probably be a long living overarching epic to collect all features of a future notification service.|
|----|

The notification service will consist of multiple micro services with different responsibilities. The services is to be loosely connected to each other through an event system like Kafka.

![highlevel](https://user-images.githubusercontent.com/2217340/200877481-fe4251de-d1c2-4938-aeb2-baf07364675e.png)

```[tasklist]
### Related epics
- [ ] #148
- [ ] https://github.com/Altinn/altinn-notifications/issues/201
- [ ] https://github.com/Altinn/altinn-notifications/issues/203
- [ ] https://github.com/Altinn/altinn-notifications/issues/204
- [ ] https://github.com/Altinn/altinn-notifications/issues/34
```

### Send SMS
This service should be able to send an SMS to a given phone number. The text must be provided in its entirety. Might require a sending window that prevents SMS from being sent during the night. 

**Possible challenges**:
- Setting who the message is from. There have in the past been limits to how many different "from names" we're allowed to use, but that limit might be related to current provider.
- Finding a service provider we can use.

### Send email
This service should be able to send an email to a given email address. The subject and body of the email must be provided in its entirety. Should support HTML emails as well as rich and plain text.

**Possible challenges**:
- Finding a suitable email server. When one email server starts producing a lot of similar emails it runs the risk of being blocked by email providers and intermediary systems. This has happened to Altinn 2 in the past.
- [DMARC](https://en.wikipedia.org/wiki/DMARC)?

### Send other
In addition to SMS and email it should be easy to expand with other message/notification channels. This could be separated into one micro services for each channel, but it might also be possible to support something like webhook and that way be able to post notifications to a long range of applications.

### Scheduling
The idea here is to allow the creator of notifications to set a date and time for when the notifications should be sent. We might want to limit how far into the future we want to allow. This could go hand in hand with reminders which are conditional notification to be sent in the future.

### Reminders
Check the state of an item and determine if recipients must be reminded about the item. The delay should be adjustable based on request input.

We would like a relatively generic way of defining a condition for when the notification should be sent. An example could be that an Altinn 3 instance created by an application owner have status "not read". A generic system could allow the reminders service too look up "any" external resource. Here we could define a contract for how the request will look and what the service expect to get in response.

### Recipient identification
Dynamically identify correct recipient "address" (email, phone number, etc..) based on multiple possible rules and relationship between recipients and the item owner. We should look into making it easy to add new "rules" without the need to change the service contract.

**Examples**:
- Anyone with role DAGL for given organisation.
- KRR registered email address for given person.
- KRR registered email address for trusted third party. E.g: Guardian.
- etc

### Other feature requests
- A mechanism that can collect multiple notifications and send them all in one message instead of as individual messages. This should be an option for people in jobs that are frequent recipients of notifications. This feature should go through legal to see if it's something we can provide. 
- Portal notifications. Have the ability to provide a list of notifications relevant for the current user. In combination with a system to collect notifications? 

## Altinn 2 features
| THIS SECTION IS UNDER CONSTRUCTION |
|----|

| Keyword | Altinn 2 feature | Altinn 3 plans |
|-----|----|----|
| ShipmentDateTime | The ability to place a notification order that will trigger a notification some time in the future. The shipment date would basically postpone the time for when the notification is sent to a the datE time value indicated in the request. | This will be implemented in the new solution as well. | 
| FromAddress | The ability to set the from address when sending a notification as an email. | This is a feature that will be implemented in Altinn 3.|
| UseServiceOwnerShort NameAsSenderOfSms | The ability to set the from field in an SMS to a value stored on the service owner in Altinn 2 | This feature will be continued in Altinn 3. Ideally as a text field where service owner can set the value for each notification order. |
| Recipient address | In Altinn 2 it is optional to provide the email address or phone number of the recipient of the notification. The alternative to providing actual contact information is to provide a recipient. More about this later. | Altinn 3 will have multiple different API endpoints designed for specific purposes. The basic notification mechanism will require an email address or phone number. More information abot support for more advanced profile/register lookup later. |
| IsReservable | The ability to ignore reservations. A service owner can choose to ignore a persons reservation choice in KRR. | UP FOR DISCUSSION. It's not been decided whether Altinn 3 will check reservation or assume that the service owner checks before calling the notification service. |
| Language | Closely related to notification type, this simply just filter texts on the notification type.  | Notification types will not be implemented like today if at all. The plan is to start without. This means the service owner will need to provide the texts the way they want them to be in the final notification. The service owner will need to know the user preferred language before placing the notification order.|
| NotificationType | The notification service requires pre registered notification types. Notification types contains the texts to use in an SMS or Email in different languages. The texts can have **Macros** and **TextTokens**. More about those later. | Notification types will not be implemented like today if at all. The plan is to start without. This means the service owner will need to provide the texts the way they want them to be in the final notification. |
| Reminders | The ability to conditionally determine if a notification is needed or not. Implemented as a hierarchy of notification types. Notifications with a notification type with a parent notification type would be assumed to be a reminder. A reminder might not be sent if the condition of the  reportee element is false. An example would be if a correspondence is read, the condition becomes false and the reminder is not sent. Limited to Altinn 2 correspondence and forms. | This will be implemented by allowing the service owner to provide the condition as a form of expression: `altinn3.instance is unread`. Altinn 3 will start by implementing support for status checking of instances in Altinn 3, but can expand with other services in the future. One such possible target is Dialogporten. |
| Text macros | Macros is one of two mechanisms to allow for dynamic changes in notification texts. The templating system (Notification types) meant that the texts were defined before it was known who they would be sent to. Macros are searched for and replaced in texts; E.g. &#36;reporteeFirstName&#36; would be replaced with the name of the person targeted by the notification. All keywords are meant to find values in register and profile information about the reportee. | Currently assumed to be unnecessary with the removal of templates. Texts are defined for each notification instead of predefined. There should be limited need to make dynamic changes to the texts. |
| Text tokens | Tokens are the second mechanism that allow for dynamic changes in notification texts. A notification text could have x number of tokens which would be replaced by text given when placing the notification request. | Currently also assumed to be unnecessary for the same reason as for text macros. |

### Recipients - advanced transport type and address lookup

A very important difference between the Altinn 2 and Altinn 3 solution will be to what degree we will allow the notification system to perform tasks not strictly related to notifications. The goal is to implement everything the service owner will need, but it isn't given that all of it will be implemented as a part of notification. One such case is dynamically identifying receiver addresses based on a national identity or organisation number. Those features might instead become a part of a new profile application.

This will affect everything related to: reportee number, transport types, roles and service details. All of those input parameters to Altinn 2 notification where input values to a system trying to find correct email address or phone number for the recipient of the notification.

| Keyword | Altinn 2 feature | Altinn 3 plans |
|-----|----|----|
| ReporteeNumber | Notification would based on the reportee number value try to find one or more email addresses or phone number to use as recipients of a notification. If the recipient is a person, notification will look in KRR for an address. If the recipient is an organisation, notification will look in ER and the personal contact register\* | We want this feature to be continued in Altinn 3, but we're not sure in what sub system of Altinn 3 it will be natural to have a feature like this. |

\* The personal contact register is a part of profile in Altinn 2. It allows a user to register their own contact information for when they represent an organisation. "Your contact information for the organisation."

- ReceiverEndPoints (list)
    - TransportType
        - SMS
        - Email
        - IM
        - Both
        - SMSPreferred
        - EmailPreferred
        - PriorityEmailSMSReminder
    - ReceiverAddress
- ReporteeNumber
- Roles (list)
    - RoleName
- Service 
    - ServiceCode
    - ServiceEdition

**Comment by olebhansen** [comment]
Closing to clean up.
